Feature/multiindex

mreijnde opened the pull request:
work in progress.

mreijnde commented:
I have rewritten dataframe horzcat() and vertcat() such that the main combination code resides in the Index class, and is compatible with MultiIndex. The Index class has a union_ function that has the logic to merge multiple Index objects and output position indexes with the mapping to this new index. These position indexes can than be used in turn to easily combine data frame multiple dataframes.

The original code has some strict limitations on what dataframes are allowed to be combined. For example, vertcat requires that rows are unique. I copied these restrictions, to get the same error behavior as before (with some minor exception). I think it would make sense to lift these restrictions further. The new code can handle all cases. A warning is given if dataframes have overlapping values.

mreijnde commented on +frames/private/nanunique.m:
```diff
@@ -0,0 +1,35 @@
+function [C,ia,ic]= nanunique(x, varargin)
+% modified unique function that threats NaN values not as distinct values.
```
On this moment this function is only used on the 'uniqueind' variable, which is a position index, and therefor always numerical.

mreijnde commented on +frames/MultiIndex.m:
```diff
@@ -0,0 +1,681 @@
+classdef MultiIndex < frames.Index
```
- A more abstract parent class maybe an good option later, not sure if that is now worth the effort.
- The MultiIndex class can also use unique and sorted. It will than look at the full rows (consisting of multiple dimensions at once)
- On the end it should work with all the computations. It should behave the same way (except maybe some places where I like to break the current behavior eg with auto alignment).  Most of the look and feel should be the same, only for example  .value will be slightly different.

mreijnde commented on +frames/Index.m:
```diff
@@ -83,36 +95,42 @@
+        function obj = set.value_(obj,value)
+            % set lowlevel value, and trigger cache update
             obj.value_ = value;
+            obj = obj.recalc_unique_cache();
```
That was also my original idea, but I couldn't find a way to get it to work due to matlab limitations. Setters are not allowed to change other properties, except the value itself. Eg, the set.value is not allowed to change an other property valueChanged.

I tried some combinations, but the changedValue was not updated. This may be related that it is a value based class, that makes updating the object more difficult.

mreijnde commented:
Locally I do not get any errors on the checks in the testsuite. It is not clear for me yet why it fails here.

benjamingaudin commented:
My bad, it's a bug on develop. Please merge the new version of develop.

mreijnde commented on +frames/Groups.m:
```diff
@@ -110,6 +110,7 @@
                 data = g.data(:)';
                 missingData = ismissing(data);
                 if obj.isColumnGroups, obj.groupless = g.columns(missingData); else, obj.groupless = g.rows(missingData)'; end
+                if numel(obj.groupless)==0, obj.groupless= []; end % fix for empty string array not size [0,0]
```
I don't understand yet what behavior changed that this line is now needed. Without this line groupless can be an empty string array with size [0 1], and some horzcat will fail later in the code because of that.

benjamingaudin commented on +frames/DataFrame.m:
```diff
@@ -1432,6 +1985,12 @@ function columnsValidation(obj,value)
             df.rows_.requireUnique = nameValue.Unique;
             df.rows_.name = string(t.Properties.DimensionNames{1});
         end
+        
```
add a `restoreDefaultSettings` function?

mreijnde replied:
added function to restore defaultsettings to 'out-of-the-box' default. It does not change settings existing DataFrame objects.

benjamingaudin commented on +frames/DataFrameSettings.m:
```diff
@@ -0,0 +1,21 @@
+classdef (HandleCompatible) DataFrameSettings
```
Can we put this and its child to +internal? Or when do we need to use this class and its constructor explicitly?

mreijnde replied:
ok, moved

benjamingaudin commented on +frames/Index.m:
```diff
@@ -167,42 +185,78 @@
-            %    - userCall       (logical):    perform full validation of selector
+            %    - positionIndex  (logical):    selector is position index instead of value index (default false)
+            %    - userCall       (logical):    perform full validation of selector (default true)
+            %    - allowMissing   (logical):    allow selectors with no matches (default error)
```
default false

mreijnde replied:
what do you mean?

mreijnde replied:
fixed

benjamingaudin commented on +frames/Index.m:
```diff
@@ -234,47 +302,345 @@
+            %    - obj1,obj2:    Index objects to be aligned
+            %
+            %    - method: (string enum) select method
+            %           "strict": both need to have same values (else error thrown)
```
I think that the method should return an error with duplicates, unless they are exactly equal (in the same order). Otherwise you can get unexpected behavior with elementWise operations and aligment.

Pandas duplicates the index for elementWise operations (if you have idx1 = [3,3], idx2=[3,3] -> idxOut = [3,3,3,3]. And fails when trying to concatenate DataFrames with duplicated indices (`InvalidIndexError: Reindexing only valid with uniquely valued Index objects`).

mreijnde replied:
Good remark, it seems indeed that the alignment behavior with duplicates is not implemented correctly. 

If the indices are exactly the same, they are just 1:1 mapped.  This is same behavior as the current frames. But when the indices are not completely equal, the result does not align the duplicates currently. The are just ignored (made NaN).

I agree that a solution that assumes the duplicates are ordered the same way (when index is not exactly the same)  is dangerous. Expanding (like pandas) to all the combinations is I think  not desirable for most use-cases, and not consistent with the current behavior (when indices are equal). Throwing an error is indeed most logical solution.

Remark: 
The auto-expansion (in case of elementwise operations with MultiIndex with different dimension names) does expands to all the combinations including duplicates.

mreijnde replied:
To get the required behavior I splitted the behavior for duplicates from the alignMethod setting. 

duplicateOptions determines behavior with duplicates in the alignment between 2 dataframes, alignMethod determines what unique values are kept in the indices from both dataframes.


         alignMethod (1,1) string {mustBeMember(alignMethod,["strict", "inner", "left", "full","strictunique"])} = "strict"
                             % defines how indices are aligned in math operation between DataFrames:
                             % -  "strict":  both need to have same indices (else error thrown)
                             % -  "inner":   remove index values that are not common in both
                             % -  "left":    keep indices as in left DataFrame in operation (default)            
                             % -  "full":    keep all indices values (allow missing values in indices in both)
                       
        duplicateOption (1,1) string ...
                        {mustBeMember(duplicateOption,["none", "unique", "duplicates","duplicatesstrict","expand"])} ...
                        = "duplicatesstrict"
                             % defines how duplicate values in indices are handled in math operation on DataFrames:                             
                             %  - "none":             no operations with duplicates allowed
                             %  - "unique":           removes duplicates, only use first occurrence of duplicate value
                             %  - "duplicatestrict":  only allow duplicates in case of equal indices
                             %  - "duplicates":       align duplicates in order of occurrence between indices
                             %                        (first occurrence duplicate value in one index is matched to 
                             %                         1st occurence in other, 2nd occurence to 2nd etc)
                             %  - "expand":           create all combinations of matching duplicates between indices
                             %                        in `outputs


The 'duplicatestrict' default duplicationOption now make sure that duplicates are only allowed if the indices are exactly the same. An expand option is now also available to get the behavior as in pandas.

benjamingaudin commented on +frames/MultiIndex.m:
```diff
@@ -0,0 +1,974 @@
+    %
+    %      
+    methods
+        function obj = MultiIndex(value, nameValue)
```
`frames.MultiIndex({[1 1 2 2],["a" "b" "a" "b"]})` should not give a warning. The underlying Indexes have duplicates, but not the MultiIndex

mreijnde replied:
Agree, that is downside of reusing the Index class as sub-index in MultiIndex class. Some functionality is not wanted in that case.

I want to solve this by adding a property to Index class (eg. "noUniquenessWarnings") to prevent warning about uniqueness.

mreijnde replied:
done, implemented hidden properties in Index obj 'warningNonUnique_'. No unique warnings of Index are raised anymore when using MultiIndex.

benjamingaudin commented on +frames/MultiIndex.m:
```diff
@@ -0,0 +1,974 @@
+        end
+        
+        
+        function out = getSelector(obj,selector, positionIndex, allowedSeries, userCall, asFilter)
```
A user should not have to enter `userCall=true`. Put it at the end or use namedArguments

mreijnde replied:
Not sure if it is important, as function itself is not exposed.  Anyway, I changed parameter order to put userCall argument on the end.

benjamingaudin commented on +frames/MultiIndex.m:
```diff
@@ -0,0 +1,974 @@
+            %    - obj1,obj2: MultiIndex objects to be aligned
+            %
+            %    - method: (string enum) select method for common dimension(s)
+            %           "strict": both need to have same values (else error thrown)
```
Can we name them like SQL "strict", "inner", "left", "full"?

mreijnde replied:
done

benjamingaudin commented on +frames/MultiIndex.m:
```diff
@@ -0,0 +1,974 @@
+        end
+        
+        
+         function obj=getSubIndex_(obj,selector, dimindex)
```
Put internal methods in protected (or hidden) block

benjamingaudin commented on +frames/test_multiIndex_selection.m:
```diff
@@ -0,0 +1,96 @@
+warning('off', 'frames:Index:notUnique');
```
this script does not run

mreijnde replied:
File was added by mistake, it is already removed.

benjamingaudin commented on +frames/private/groupsummaryMatrixFast.m:
```diff
@@ -0,0 +1,211 @@
+function [B, groupID, groupCount, groupFirstInd] = groupsummaryMatrixFast(A, groupid, func, dim, funcAggrDim, apply2single, vectorize, convGroupInd)
```
Looks good but what makes it faster / what makes Matlab function slow?

mreijnde replied:
Not sure how the matlab builtin algorithm works, but it looks like it does calculate the aggregation groups again for every column. That is inefficient with larger number of columns.

Next to that this function can skip applying the aggregation function if only a single value is in the group.

Additional functionality is added to also aggregate columns and still make use of the aggregation function vectorization.

benjamingaudin commented on +frames/private/uniqueCellRows.m:
```diff
@@ -0,0 +1,33 @@
+function [rows_uniq, rows_uniqInd, val_uniq, val_ind] = uniqueCellRows(data,order)
```
This fct is not used

mreijnde replied:
indeed not used anymore after I updated the code in some iterations. Removed file.

benjamingaudin commented on scripts/example_testmultiindex.m:
```diff
@@ -0,0 +1,32 @@
+% script with some basic examples/tests for MultiIndex
```
is this script needed, given the mlx?

mreijnde replied:
It was just meant as simple example, in later phase it has to be merged with the demo mlx file (or some other demonstration script).

benjamingaudin commented on scripts/example_testmultiindex.m:
```diff
@@ -0,0 +1,32 @@
+df2 = df1.groupDuplicate(@mean)
+
+% manipulations
+df3 = df2.sum("x")                     % aggregate over (sub)dimension
```
so that removes the "x" and does not groupby the values in "x". Is that what you want?

mreijnde replied:
Yes, I think this is the right behavior. The math function sum() (and others like mean(), std(), max(),...) are aggregation functions. Before you could only aggregate with these functions over the full rows or column dimensions. Now it is expanded such that you can only aggregate over a sub-dimension like "x" in this case.

If you do the aggregation over a dimension, the groupby is done over the remaining dimensions. For example if aggregation in "x" is done, it is grouped by dimension "y" and "z". You should look at this if the data is in a (sparse) 3d cube. By using aggregation a single dimension of this cube is flattened.

This is also similar as done in for example python xarray library, which is specialized in processing of multi-dimensional datasets. The MultiIndex additions to frames also bring such features.

benjamingaudin commented on test/unit/frames/indexTest.m:
```diff
@@ -133,7 +133,7 @@ function assignmentTest(t)
         function cellForTimeForbidden(t)
             t.verifyError(@() frames.TimeIndex({'02.11.2021'}),'TimeIndex:cellstrnotsupported')
-            t.verifyError(@() frames.TimeFrame(1,{"02.11.2021"}),'TimeIndex:cellstrnotsupported') %#ok<STRSCALR>
+            %t.verifyError(@() frames.TimeFrame(1,{"02.11.2021"}),'TimeIndex:cellstrnotsupported') %#ok<STRSCALR> <== not compatible with new MultiIndex {} syntax
```
still test the error with timeframe

mreijnde replied:
enabled test with some explanation

benjamingaudin commented on +frames/Index.m:
```diff
@@ -234,44 +270,196 @@
-            obj.singleton_ = false;
-            obj.value = val;  % check if properties are respected
+            obj = obj.vertcat_(varargin{:});
+            obj.setvalue(obj.value_); % check if properties are respected            
```
instead of calling the whole `setvalue`

benjamingaudin commented on +frames/DataFrame.m:
```diff
@@ -1657,7 +1685,9 @@ function rowsColChecker(df1,df2)
 function testUniqueIndex(indexObj)
-if ~indexObj.requireUnique
+% test require unique, and ignore empty index
+if ~indexObj.requireUnique && length(indexObj.value_)>0
```
is it needed?

benjamingaudin commented on +frames/Index.m:
```diff
@@ -25,20 +25,28 @@
-    properties(Hidden,Access={?frames.TimeIndex,?frames.DataFrame})
-        value_
+    properties      %TEMPORARY FOR DEBUGGING (NO ACCESS LIMITATIONS)
+        value_               % stores values   
```
back to the block below

mreijnde replied:
done

benjamingaudin reviewed (commented):
This is excellent. Here are my first comments, let me know what you think

mreijnde commented on +frames/getPrivateFuncHandle.m:
```diff
@@ -0,0 +1,5 @@
+function fh = getPrivateFuncHandle(funcname)
```
Agree, done

mreijnde commented:
@benjamingaudin: do you know why the unit-tester does not run anymore after a commit?

benjamingaudin commented:
> @benjamingaudin: do you know why the unit-tester does not run anymore after a commit?

Git still believes there’s a conflict with develop. Maybe that’s why they don’t run

mreijnde commented:
I addressed your comments, if you have time, please have another look. I also added more testcases to check the expansion and alignment behavior.

benjamingaudin reviewed (changes requested):
Hi Merijn,

I added some minor comments; there are a few left from before. Please have a look at them too.

Before we can merge your work, can I ask you to:
  - Adapt Frames.align function with your work with alignMethod and duplicateOptions. For now, the method only expands the rows. I would be nice to have a way to provide your additional settings, and to be able to do it on columns too
  - Add a mention that the MultiIndex exists in the readme (with a short description) and add a link in the readme to the html version of your mlx file (.mlx -> save as .m -> publish), to advertise your work

benjamingaudin commented:
I think `[frames.DataFrame(1,ColSeries=true), frames.DataFrame(2,[],"col")]` should either return a frame with two columns, or an error, but not just delete the colseries, or merge the data.

mreijnde commented on +frames/DataFrame.m:
```diff
@@ -499,115 +638,140 @@
+                assert(options.duplicateOptionCols=="unique", ...
+                    'frames:DataFrame:combine:invalidColsMethod', ...
+                    "Invalid alignMethodCols option. Column index has requireUnique enabled, only 'unique' allowed.");
+                columns_requireUnique = cellfun(@(x) x.columns_.isunique() || length(x.columns_)==0, df);
```
removed check, could identify usecase why this length check was added.

mreijnde commented on +frames/MultiIndex.m:
```diff
@@ -0,0 +1,975 @@
+            %           "none":             do not allow duplicates present            
+            %           "duplicatesstrict": do not allow duplicates present, except fully equal indices (default)
+            %           "duplicates":       align duplicates in order of appearance
+            %
```
fixed

mreijnde commented:
> I think `[frames.DataFrame(1,ColSeries=true), frames.DataFrame(2,[],"col")]` should either return a frame with two columns, or an error, but not just delete the colseries, or merge the data.

I think the behavior is as expected as you concat an unique index. The singular column index maps to every column index. The dataframe values are applied in order, such that the last matching one will determine the values. The combine method has the order as input parameter.

It produces a warning: 
```
Warning: Overlapping values (with same row and column index) between different dataframes detected.
Value of last dataframe will be used. 
> In frames/DataFrame/combine (line 751)
In frames/DataFrame/vertcat (line 777) 
```

In this case:  `[ frames.DataFrame([2 2],[],["col","col2"]) frames.DataFrame(1,ColSeries=true)] `  both columns are overwritten.

mreijnde commented:
I implemented the changes as requested. I edited the readme file and added the existing multiIndex live document as html. The readability of that document can be improved, but hope it is clear enough for now.

I also modified the code to use enumeration class for the alignMethod and duplicateOption parameters, to make the code more clear.

benjamingaudin reviewed (approved):
Congrats on this beautiful work. It's all good! 

If there is no more work pending on your side, I will merge your branch and then release this version.

I just have one comment left on the `getSelector` functions if you want to address it.

mreijnde commented:
Hi Benjamin,

Thx for reviewing it. On this moment dont have concrete additions planned,
maybe in the future.

I will try to have a look tomorrow. Didnt have oppertunity yet as I was
traveling this week.

regards Merijn



On Tue, 7 Jun 2022, 12:09 Benjamin Gaudin, ***@***.***> wrote:

> ***@***.**** approved this pull request.
>
> Congrats on this beautiful work. It's all good!
>
> If there is no more work pending on your side, I will merge your branch
> and then release this version.
>
> I just have one comment left on the getSelector functions if you want to
> address it.
> ------------------------------
>
> In +frames/MultiIndex.m
> <https://github.com/benjamingaudin/Frames/pull/35#discussion_r890941525>:
>
> > +            % internal use: no validation checks, see details at getSubIndex()
> +            %
> +            % select dimensions
> +            if ~iscolon(dimindex)
> +                obj.value_ = obj.value_(dimindex);
> +            end
> +            % select rows
> +            if ~iscolon(selector)
> +                for i=1:obj.Ndim
> +                  obj.value_{i} = obj.value_{i}.getSubIndex(selector);
> +                end
> +            end
> +        end
> +
> +
> +        function out = getSelector(obj,selector, positionIndex, allowedSeries, userCall, asFilter)
>
> I don't understand what you mean by "the function is not exposed". It is
> public isn't it?
>
> I think you did not commit the change you mention.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/benjamingaudin/Frames/pull/35#pullrequestreview-997752532>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEXYXDE4UFQNLFYIBU2JTELVN4NXFANCNFSM5FKM2FZQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

mreijnde commented:
On this moment I don't have others changes pending.
